### PR TITLE
OSPlatform attributes breaking change

### LINKED ISSUE #20635: [Breaking Change] MinimumOSPlatformAttribute and RemovedInOSPlatformAttribute renamed, ObsoletedInOSPlatformAttribute removed
## MinimumOSPlatformAttribute and RemovedInOSPlatformAttribute renamed, ObsoletedInOSPlatformAttribute removed

[`MinimumOSPlatformAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.versioning.minimumosplatformattribute) was renamed to `SupportedOSPlatformAttribute`; [`RemovedInOSPlatformAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.versioning.removedinosplatformattribute) was renamed to `UnsupportedOSPlatformAttribute`; and [`ObsoletedInOSPlatformAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.versioning.obsoletedinosplatformattribute) was removed.

These changes can introduce build-time or runtime errors for projects when retargeting from .NET 5.0 Preview 8 to .NET 5.0 RC1.  The rename of `MinimumOSPlatformAttribute` is most likely to produce build-time or runtime errors in projects because the attribute is applied to platform-specific assemblies at build-time and old build artifacts will still reference the old API name.

Example build-time errors:
```
error CS0246: The type or namespace name 'MinimumOSPlatformAttribute' could not be found (are you missing a using directive or an assembly reference?)
error CS0246: The type or namespace name 'RemovedInOSPlatformAttribute' could not be found (are you missing a using directive or an assembly reference?)
error CS0246: The type or namespace name 'ObsoletedInOSPlatformAttribute' could not be found (are you missing a using directive or an assembly reference?)
```

Example runtime error:
```
Unhandled exception. System.TypeLoadException: Could not load type 'System.Runtime.Versioning.MinimumOSPlatformAttribute' from assembly 'System.Runtime, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
```

These errors are likely to be introduced for projects targeting OS specific flavors of .NET 5 through target framework monikers such as `net5.0-windows`. To resolve either error, the project must be rebuilt (clean + build) after retargeting to .NET 5.0 RC1.

### Version introduced

.NET 5.0 RC1

### Old behavior

In .NET 5.0 Preview 8, the following attributes existed within `System.Runtime.Versioning`:

1. `MinimumOSPlatformAttribute`
2. `RemovedInOSPlatformAttribute`
3. `ObsoletedInOSPlatformAttribute`

In .NET 5.0 Preview 8, when a project was targeting an OS specific flavor of .NET 5 through a target framework moniker such as `net5.0-windows`, the build would produce an assembly-level attribute using `System.Runtime.Versioning.MinimumOSPlatformAttribute`.

### New behavior

In .NET 5.0 RC1, the following changes were made to the attributes:

1. `MinimumOSPlatformAttribute` was renamed to `SupportedOSPlatformAttribute`
2. `RemovedInOSPlatformAttribute` was renamed to `UnsupportedOSPlatformAttribute`
3. `ObsoletedInOSPlatformAttribute` was removed

In .NET 5.0 RC1, when a project was targeting an OS specific flavor of .NET 5 through a target framework moniker such as `net5.0-windows`, the build will produce an assembly-level attribute using `System.Runtime.Versioning.SupportedOSPlatformAttribute`.

### Reason for change

.NET 5.0 Preview 8 introduced attributes in `System.Runtime.Versioning` with the purpose of performing [platform checks](https://github.com/dotnet/designs/blob/main/accepted/2020/platform-checks/platform-checks.md) for APIs that only work on specified platforms. The attributes are consumed by the Platform Compatibility Analyzer to produce build warnings when consuming platform-specific APIs on platforms those APIs don't support.

During .NET 5.0 RC1, an additional feature was added to the Platform Compatibility Analyzer for [platform exclusion](https://github.com/dotnet/designs/blob/main/accepted/2020/platform-exclusion/platform-exclusion.md), allowing APIs to be marked as entirely unsupported on OS platforms. That feature prompted changes to the attributes, using API names more suitable for scenarios.

The following API changes were made in `System.Runtime.Versioning` to support this new feature:

1. `MinimumOSPlatformAttribute` renamed to `SupportedOSPlatformAttribute`, to support platform-specific API annotations
2. `RemovedInOSPlatformAttribute` renamed to `UnsupportedOSPlatformAttribute`, to support platform exclusion annotations
3. `ObsoletedInOSPlatformAttribute` was removed, as this annotation is not needed at this time

### Recommended action

1. After retargeting the project to .NET 5.0 RC1, rebuild the project (or perform clean + build)
2. Update any references of `MinimumOSPlatformAttribute` to `SupportedOSPlatformAttribute`
3. Update any references of `RemovedInOSPlatformAttribute` to `UnsupportedOSPlatformAttribute`
4. Remove any references to `ObsoletedInOSPlatformAttribute`

### Category

- [ ] ASP.NET Core
- [ ] C#
- [x] Code analysis
- [x] Core .NET libraries
- [ ] Cryptography
- [ ] Data
- [ ] Debugger
- [ ] Deployment
- [ ] Globalization
- [ ] Interop
- [ ] JIT
- [ ] LINQ
- [ ] Managed Extensibility Framework (MEF)
- [ ] MSBuild
- [ ] Networking
- [ ] Printing
- [ ] Security
- [ ] Serialization
- [ ] Visual Basic
- [ ] Windows Forms
- [ ] Windows Presentation Foundation (WPF)
- [ ] XML, XSLT

### Affected APIs

* [`MinimumOSPlatformAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.versioning.minimumosplatformattribute)
* [`RemovedInOSPlatformAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.versioning.removedinosplatformattribute)
* [`ObsoletedInOSPlatformAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.versioning.obsoletedinosplatformattribute)

---
#### Issue metadata

* Issue type: breaking-change

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #20635.

[Preview link](https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/corefx?branch=pr-en-us-20659#osplatform-attributes-renamed-or-removed).

(The warnings will go away once we merge the RC1 API doc updates.)

### CODE REVIEW

**Review by jeffhandley — APPROVED** [review]
This is great! Thanks for doctoring up what I wrote to make it much nicer!

**Comment by gewarren** [comment]
> This is great! Thanks for doctoring up what I wrote to make it much nicer!

You wrote it up beautifully. I hardly had to change anything!

File: includes/core-changes/corefx/5.0/os-platform-attributes-renamed.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+Core .NET libraries
+
+#### Affected APIs
+
+- <xref:System.Runtime.Versioning.MinimumOSPlatformAttribute?displayProperty=fullName>
+- <xref:System.Runtime.Versioning.ObsoletedInOSPlatformAttribute?displayProperty=fullName>
+- <xref:System.Runtime.Versioning.RemovedInOSPlatformAttribute?displayProperty=fullName>
+
+<!--
```
Comment (BillWagner) [inline includes/core-changes/corefx/5.0/os-platform-attributes-renamed.md:65]: Any reason to leave the commented out links here?

**Review by BillWagner — APPROVED** [review]
This also LGTM @gewarren. I did have one question.

Once the dependencies are merged, you can :shipit:

File: includes/core-changes/corefx/5.0/os-platform-attributes-renamed.md (hunk 1 above)
Comment (gewarren) [reply includes/core-changes/corefx/5.0/os-platform-attributes-renamed.md:65]: Yes, they're used by the API analyzer.

**Comment by gewarren** [comment]
Close/reopen for fresh build.

**Pull request merged**